scilpy/tractanalysis/fibertube_scoring.py

```diff
@@ -14,59 +15,142 @@
+        Number of samples to be created along a single axis of a voxel. The
+        total number of samples in the voxel will be this number cubed.
+    verbose: bool
+        Whether the function and sub-functions should be verbose.
 
     Returns
     -------
-    mean_density: float
-        Per-voxel spatial density, averaged for the whole tractogram.
+    density_grid: ndarray
+        Per-voxel volumetric density of fibertubes as a 3D image.
+    density_flat: list
+        Per-voxel volumetric density of fibertubes as a list, containing only
+        the voxels that were valid in the binary mask (i.e. that contained
+        fibertubes). This is useful for calculating measurements on the
+        various density values, like mean, median, etc.
+    collision_grid: ndarray
+        Per-voxel density of fibertube collisions.
+    collision_flat: list
+        Per-voxel density of fibertubes collisions as a list, containing only
+        the voxels that were valid in the binary mask (i.e. that contained
+        fibertubes). This is useful for calculating measurements on the
+        various density values, like mean, median, etc.
     """
+    if "diameters" not in sft.data_per_streamline:
+        raise ValueError('No diameter found as data_per_streamline '
+                         'in the provided tractogram')
     diameters = np.reshape(sft.data_per_streamline['diameters'],
                            len(sft.streamlines))
-    mean_diameter = np.mean(diameters)
-
-    mean_segment_lengths = []
-    for streamline in sft.streamlines:
-        mean_segment_lengths.append(
-            np.mean(np.linalg.norm(streamline[1:] - streamline[:-1], axis=-1)))
-    mean_segment_length = np.mean(mean_segment_lengths)
-    # Computing mean tube density per voxel.
+    max_diameter = np.max(diameters)
+
+    # Everything will be in vox for simplicity.
     sft.to_vox()
+    # Building a binary mask using TODI
     # Because compute_todi expects streamline points (in voxel coordinates)
     # to be in the range [0, size] rather than [-0.5, size - 0.5], we shift
     # the voxel origin to corner.
     sft.to_corner()
-
-    # Computing TDI
     _, data_shape, _, _ = sft.space_attributes
     todi_obj = TrackOrientationDensityImaging(tuple(data_shape))
-    todi_obj.compute_todi(sft.streamlines)
-    img = todi_obj.get_tdi()
-    img = todi_obj.reshape_to_3d(img)
-
-    nb_voxels_nonzero = np.count_nonzero(img)
-    sum = np.sum(img, axis=-1)
-    sum = np.sum(sum, axis=-1)
-    sum = np.sum(sum, axis=-1)
-
-    mean_seg_volume = np.pi * ((mean_diameter/2) ** 2) * mean_segment_length
-
-    mean_seg_count = sum / nb_voxels_nonzero
-    mean_volume = mean_seg_count * mean_seg_volume
-    mean_density = mean_volume / (sft.voxel_sizes[0] *
-                                  sft.voxel_sizes[1] *
-                                  sft.voxel_sizes[2])
-
-    return mean_density
+    todi_obj.compute_todi(sft.streamlines, False)
```

Comment:
TODI only for getting the mask? This is overkill and I think it might skip voxels if there are not points in each voxel. I think you could use `streamlines.tractogram.uncompress` fort this purpose.